test/integration/controller_binding_test.go

```diff
@@ -0,0 +1,532 @@
+						t.Fatalf("error waiting for binding condition: %v\n"+"expecting: %+v\n"+"last seen: %+v", err, condition, cond)
+					}
+				}
+			})
+		})
+	}
+}
+
+// TestCreateServiceBindingInvalidInstanceFailure try to bind to invalid service instance names
+func TestCreateServiceBindingInvalidInstanceFailure(t *testing.T) {
+	cases := []struct {
+		name         string
+		instanceName *string
+	}{
+		{
+			name:         "invalid service instance name",
+			instanceName: strPtr(""),
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ct := &controllerTest{
+				t:        t,
+				broker:   getTestBroker(),
+				instance: getTestInstance(),
+			}
+			ct.run(func(ct *controllerTest) {
+				binding := getTestBinding()
+				if tc.instanceName != nil {
+					binding.Spec.ServiceInstanceRef.Name = *tc.instanceName
+				}
+
+				if _, err := ct.client.ServiceBindings(binding.Namespace).Create(binding); err == nil {
+					t.Fatalf("expected binding to fail to be created due to invalid parameters")
+				}
+			})
+		})
+	}
+}
+
+// TestCreateServiceBindingInvalidInstance try to bind to invalid service instance names
+func TestCreateServiceBindingInvalidInstance(t *testing.T) {
+	cases := []struct {
+		name         string
+		instanceName *string
+		condition    v1beta1.ServiceBindingCondition
+	}{
+		{
+			name:         "non-existent service instance name",
+			instanceName: strPtr("nothereinstance"),
+
+			condition: v1beta1.ServiceBindingCondition{
+				Type:   v1beta1.ServiceBindingConditionReady,
+				Status: v1beta1.ConditionFalse,
+				Reason: "ReferencesNonexistentInstance",
+			},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ct := &controllerTest{
+				t:        t,
+				broker:   getTestBroker(),
+				instance: getTestInstance(),
+				binding: func() *v1beta1.ServiceBinding {
+					b := getTestBinding()
+					if tc.instanceName != nil {
+						b.Spec.ServiceInstanceRef.Name = *tc.instanceName
+					}
+					return b
+				}(),
+				skipVerifyingBindingSuccess: true,
+			}
+			ct.run(func(ct *controllerTest) {
+				if cond, err := util.WaitForBindingCondition(ct.client, testNamespace, testBindingName, tc.condition); err != nil {
+					t.Fatalf("error waiting for binding condition: %v\n"+"expecting: %+v\n"+"last seen: %+v", err, tc.condition, cond)
+				}
+			})
+		})
+	}
+}
+
+// TODO: this one is still broken:
+// E1116 16:10:21.295296    8302 controller_instance.go:1844] ServiceInstance "test-namespace/test-instance": Failed to update status: ServiceInstance.servicecatalog.k8s.io "test-instance" is invalid: status.currentOperation: Forbidden: currentOperation must not be present when reconciledGeneration and generation are equal
+// TestCreateServiceBindingAsyncServiceInstance try to bind to a in progress service instance.
+//func TestCreateServiceBindingAsync(t *testing.T) {
```

Comment:
Let's remove this function from this PR rather than introduce a bunch of commented-out code. You can add the test as a comment in #1587, so that the test gets added as part of resolving the bug that is blocking this test.